### PR TITLE
Add resource middleware on actix-web-codegen

### LINKED ISSUE #1466: Allow to specify a middleware with codegen
In my project, I would like to apply an authentication middleware for each resource I have. I've searched on the web, examples, and issues, but I didn't find a way to do it simply. 

In my case, I use an enum `Authentication`, that implements `Transform` to `Service` (a middleware).

```rust
enum Authentication {
    Public,
    Private,
    Admin,
}
```
Here you have the complete code: https://github.com/meilisearch/MeiliSearch/blob/6aceef1ea0d87c0e01e8ff866239bb5738f6525c/meilisearch-http/src/helpers/authentication.rs

What I would like to do is to apply this middleware directly on the proc_macro instead of writing an ultra verbose code for each of my 48 routes. And I love the codegen crate!

And examples of what I would like at the end. 

```rust
#[get("/test/wrap", wrap = "Authentication::Public")]
async fn get_warp(_: Path<String>) -> impl Responder {
    HttpResponse::Ok()
}
```

I already tested the fix on actix-web-codegen and will do a PR on actix-web. I will probably create a temporary crate that takes up and improves the actix-web-codegen crate.

### PR DESCRIPTION
**Pull request by qdequele**
fix https://github.com/actix/actix-web/issues/1466

### CODE REVIEW

**Comment by fafhrd91** [comment]
How is it breaking change?

**Comment by qdequele** [comment]
I only add a feature, I haven't touched the other features at all. I use it in my project with perfection! An example: https://github.com/meilisearch/MeiliSearch/blob/tide-to-actix-web/meilisearch-http/src/routes/document.rs

**Comment by cdbattags** [comment]
"resource" is the correct spelling, not "ressource" just a heads up

**Comment by qdequele** [comment]
> "resource" is the correct spelling, not "ressource" just a heads up

Fixed

**Comment by qdequele** [comment]
Do you want me to add more tests?

File: actix-web-codegen/src/lib.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@
 //!
 //! - `"path"` - Raw literal string with path for which to register handle. Mandatory.
 //! - `guard="function_name"` - Registers function as guard using `actix_web::guard::fn_guard`
+//! - `wrap=Middleware` - Registers a resource middleware.
```
Comment (JohnTitor) [inline actix-web-codegen/src/lib.rs]:
```suggestion
//! - `wrap="Middleware"` - Registers a resource middleware.
```

File: actix-web-codegen/src/lib.rs [hunk 2]
```diff
@@ -54,6 +55,7 @@ use proc_macro::TokenStream;
 ///
 /// - `"path"` - Raw literal string with path for which to register handler. Mandatory.
 /// - `guard="function_name"` - Registers function as guard using `actix_web::guard::fn_guard`
+/// - `wrap=Middleware` - Registers a resource middleware.
```
Comment (JohnTitor) [inline actix-web-codegen/src/lib.rs]:
```suggestion
/// - `wrap="Middleware"` - Registers a resource middleware.
```

File: actix-web-codegen/src/route.rs [hunk 3]
```diff
@@ -85,10 +87,19 @@ impl Args {
                                 "Attribute guard expects literal string!",
                             ));
                         }
+                    } else if nv.path.is_ident("wrap") {
+                        if let syn::Lit::Str(lit) = nv.lit {
+                            wrappers.push(lit.parse()?);
+                        } else {
+                            return Err(syn::Error::new_spanned(
+                                nv.lit,
+                                "Attribute guard expects type",
```
Comment (JohnTitor) [inline actix-web-codegen/src/route.rs]: I guess we could improve the wording though.

```suggestion
                                "Attribute wrap expects type",
```

File: actix-web-codegen/tests/test_macro.rs [hunk 4]
```diff
@@ -73,6 +78,65 @@ async fn get_param_test(_: Path<String>) -> impl Responder {
...
+            let header_name = HeaderName::from_lowercase(b"custom-header").unwrap();
+            let header_value = HeaderValue::from_str("hello").unwrap();
+            headers.insert(header_name, header_value);
+            Ok(res)
+        })
+    }
+}
+
+#[get("/test/wrap", wrap = "ChangeStatusCode")]
+async fn get_warp(_: Path<String>) -> impl Responder {
```
Comment (JohnTitor) [inline actix-web-codegen/tests/test_macro.rs]:
```suggestion
async fn get_wrap(_: Path<String>) -> impl Responder {
```

File: actix-web-codegen/tests/test_macro.rs [hunk 5]
```diff
@@ -155,3 +219,15 @@ async fn test_auto_async() {
     let response = request.send().await.unwrap();
     assert!(response.status().is_success());
 }
+
+#[actix_rt::test]
+async fn test_wrap() {
+    let srv = test::start(|| {
+        App::new()
+            .service(get_warp)
```
Comment (JohnTitor) [inline actix-web-codegen/tests/test_macro.rs]:
```suggestion
            .service(get_wrap)
```

**Review by JohnTitor — COMMENTED** [review]
Looks good! Left some minor nits.

**Comment by JohnTitor** [comment]
> How is it breaking change?

Nah, that's my bad :P 

> Do you want me to add more tests?

The current is enough for me. I think we could use trybuild test here also but it's a follow-up task.

**Review by JohnTitor — APPROVED** [review]
Thanks!

**Pull request merged**